lighthouse-core/test/audits/legacy-javascript-test.js

```diff
@@ -0,0 +1,166 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const assert = require('assert');
+const LegacyJavascript = require('../../audits/legacy-javascript.js');
+const networkRecordsToDevtoolsLog = require('../network-records-to-devtools-log.js');
+
+/**
+ * @param {Array<{url: string, code: string}>} scripts
+ * @return {LH.Artifacts}
+ */
+const createArtifacts = (scripts) => {
+  const networkRecords = scripts.map(({url}, index) => ({
+    requestId: String(index),
+    url,
+  }));
+  return {
+    devtoolsLogs: {defaultPass: networkRecordsToDevtoolsLog(networkRecords)},
+    ScriptElements: scripts.reduce((acc, {code}, index) => {
+      acc[String(index)] = {
+        content: code,
+        requestId: String(index),
+      };
+      return acc;
+    }, {}),
+  };
+};
+
+/* eslint-env jest */
+describe('LegacyJavaScript audit', () => {
+  it('passes code with no polyfills', async () => {
+    const artifacts = createArtifacts([
+      {
+        code: 'var message = "hello world"; console.log(message);',
+        url: 'https://www.example.com/a.js',
+      },
+      {
+        code: 'SomeGlobal = function() {}',
+        url: 'https://www.example.com/a.js',
+      },
+      {
+        code: 'SomeClass.prototype.someFn = function() {}',
+        url: 'https://www.example.com/a.js',
+      },
+      {
+        code: 'Object.defineProperty(SomeClass.prototype, "someFn", function() {})',
+        url: 'https://www.example.com/a.js',
+      },
+    ]);
+    const result = await LegacyJavascript.audit(artifacts, {computedCache: new Map()});
+    assert.equal(result.score, 1);
+    assert.equal(result.numericValue, 0);
+  });
+
+  it('fails code with a legacy polyfill', async () => {
+    const artifacts = createArtifacts([
+      {
+        code: 'String.prototype.repeat = function() {}',
+        url: 'https://www.example.com/a.js',
+      },
+    ]);
+    const result = await LegacyJavascript.audit(artifacts, {computedCache: new Map()});
+    assert.equal(result.score, 0);
+    assert.equal(result.numericValue, 1);
+    assert.equal(result.details.items[0].description, 'String.prototype.repeat');
+  });
+
+  it('fails code with multiple legacy polyfills', async () => {
+    const artifacts = createArtifacts([
+      {
+        code: 'String.prototype.repeat = function() {}; String.prototype.includes = function() {}',
+        url: 'https://www.example.com/a.js',
+      },
+    ]);
+    const result = await LegacyJavascript.audit(artifacts, {computedCache: new Map()});
+    assert.equal(result.score, 0);
+    assert.equal(result.numericValue, 2);
+  });
+
+  it('counts multiple of the same polyfill from the same script only once', async () => {
```

Comment:
what's the motivation here? shouldn't it be flagged if they're shipping duplicate polyfills in the same bundle?

Reply:
this bit:

```
// Don't report more than one instance of a pattern for this code.
      // Would result in multiple matches for the same pattern, ex: if both '='
      // and 'Object.defineProperty' are used conditionally based on feature detection.
      // Would also result in many matches for transform patterns.
```

Can't distinguish between multiple polyfills (somehow) making their way in and just the one polyfill doing their polyfilling in multiple ways. At least, not without a more complicated matching process.